File: community-ai/packages/protocol/src/capability.ts
import { z } from "zod";

/**
 * A node advertises what it can contribute *right now*, not what hardware it owns.
 *
 * The coordinator never branches on "is this an RTX 4090". It asks the profile
 * for available capacity, memory headroom and link quality. That keeps NVIDIA,
 * AMD, Intel, Apple Silicon and CPU-only nodes on one abstraction.
 */

export const NodeKind = z.enum([
  /** Primary compute. Native worker process on a desktop/laptop. */
  "desktop-worker",
  /** Optional, experimental compute. Browser tab with WebGPU. May vanish at any time. */
  "browser-contributor",
  /** Submits work, contributes nothing. */
  "client",
]);
export type NodeKind = z.infer<typeof NodeKind>;

export const UserActivity = z.enum(["idle", "light", "active", "busy"]);
export type UserActivity = z.infer<typeof UserActivity>;

export const ThermalState = z.enum(["normal", "warm", "hot", "critical", "unknown"]);
export type ThermalState = z.infer<typeof ThermalState>;

export const AcceleratorBackend = z.enum([
  "cuda",
  "vulkan",
  "metal",
  "webgpu",
  "cpu",
  "none",
]);
export type AcceleratorBackend = z.infer<typeof AcceleratorBackend>;

export const CpuProfile = z.object({
  model: z.string(),
  cores: z.number().int().nonnegative(),
  /** Fraction of the CPU currently free, 0..1 */
  available: z.number().min(0).max(1),
});
export type CpuProfile = z.infer<typeof CpuProfile>;

export const GpuProfile = z.object({
  vendor: z.string(),
  model: z.string(),
  /** Megabytes of dedicated video memory, 0 when unknown or shared. */
  vram: z.number().nonnegative(),
  /** Fraction of the accelerator currently free, 0..1 */
  available: z.number().min(0).max(1),
  backend: AcceleratorBackend,
});
export type GpuProfile = z.infer<typeof GpuProfile>;

export const MemoryProfile = z.object({
  total: z.number().nonnegative(),
  available: z.number().nonnegative(),
});
export type MemoryProfile = z.infer<typeof MemoryProfile>;

export const NetworkProfile = z.object({
  /** Round-trip to the coordinator in milliseconds. */
  latency: z.number().nonnegative(),
  bandwidthMbps: z.number().nonnegative(),
  jitter: z.number().nonnegative(),
});
export type NetworkProfile = z.infer<typeof NetworkProfile>;

export const UserState = z.object({
  activity: UserActivity,
  thermalState: ThermalState,
  onBattery: z.boolean(),
  batteryPct: z.number().min(0).max(100).nullable(),
});
export type UserState = z.infer<typeof UserState>;

/**
 * What a node offers to a *layer-split* pipeline, which is a different
 * contribution from running a whole model. A node with 6 GB free can host a
 * slice of a 40 GB model even though it could never load one alone, so this is
 * advertised separately from the whole-model memory in `usableMemoryMB`.
 */
export const RpcProfile = z.object({
  /** `host:port` peers dial to place layers here. Null when not serving. */
  endpoint: z.string().nullable(),
  /** Free memory across every device this node exposes, as llama.cpp sees it. */
  offeredMemoryMB: z.number().nonnegative().default(0),
  devices: z
    .array(
      z.object({
        name: z.string(),
        description: z.string(),
        totalMB: z.number().nonnegative(),
        freeMB: z.number().nonnegative(),
      })
    )
    .default([]),
  /**
   * Whether this node can act as pipeline head. The head needs the full weight
   * file on disk (not in memory) plus a good link to every member, so a laptop
   * on a slow uplink should stay a member even when it has spare compute.
   */
  canHead: z.boolean().default(false),
  /** Build tag; peers must match exactly or the RPC handshake fails. */
  build: z.string().default(""),
});
export type RpcProfile = z.infer<typeof RpcProfile>;

export const RuntimeProfile = z.object({
  /** Which execution engine sits behind the AI Runtime interface. */
  engine: z.enum(["node-llama-cpp", "llama.cpp-rpc", "webllm", "none"]),
  ready: z.boolean(),
  /** Models already resident on this node, so the scheduler can prefer warm nodes. */
  loadedModels: z.array(z.string()).default([]),
  /** Model ids this node is willing and able to load. */
  supportedModels: z.array(z.string()).default([]),
  /** Absent on browser contributors, which cannot expose devices over RPC. */
  rpc: RpcProfile.nullable().default(null),
});
export type RuntimeProfile = z.infer<typeof RuntimeProfile>;

export const CapabilityProfile = z.object({
  nodeId: z.string().min(1),
  label: z.string().min(1),
  kind: NodeKind,
  platform: z.object({
    os: z.string(),
    arch: z.string(),
    version: z.string().default(""),
  }),
  cpu: CpuProfile,
  gpu: GpuProfile.nullable(),
  memory: MemoryProfile,
  network: NetworkProfile,
  userState: UserState,
  runtime: RuntimeProfile,
});
export type CapabilityProfile = z.infer<typeof CapabilityProfile>;

/**
 * The Resource Governor lives inside the worker and is the single authority on
 * how much of the owner's machine may be used. The coordinator decides WHAT a
 * node should do; this report states HOW MUCH the node will give.
 */
export const GovernorState = z.enum([
  "idle",
  "available",
  "contributing",
  "throttling",
  "paused",
  "resuming",
]);
export type GovernorState = z.infer<typeof GovernorState>;

export const GovernorReport = z.object({
  state: GovernorState,
  /** Fraction of spare resources the owner's machine is offering, 0..1 */
  capacity: z.number().min(0).max(1),
  maxConcurrentTasks: z.number().int().nonnegative(),
  /** Hard ceiling on model weights this node will hold, in MB. */
  maxModelMemoryMB: z.number().nonnegative(),
  /** Human-readable causes, surfaced in the worker CLI and the web dashboard. */
  reasons: z.array(z.string()).default([]),
  /** Owner pressed pause. Never overridden by the coordinator. */
  manualPause: z.boolean().default(false),
});
export type GovernorReport = z.infer<typeof GovernorReport>;

/** Rolling, measured performance. Not a spec sheet estimate. */
export const NodeMetrics = z.object({
  tokensPerSecond: z.number().nullable(),
  ttftMs: z.number().nullable(),
  tasksCompleted: z.number().int().nonnegative(),
  tasksFailed: z.number().int().nonnegative(),
  samples: z.number().int().nonnegative(),
});
export type NodeMetrics = z.infer<typeof NodeMetrics>;

export function emptyMetrics(): NodeMetrics {
  return {
    tokensPerSecond: null,
    ttftMs: null,
    tasksCompleted: 0,
    tasksFailed: 0,
    samples: 0,
  };
}

/**
 * Memory a node may actually devote to weights right now.
 * VRAM is first class; system RAM is discounted because host-side offload is slow.
 */
export function usableMemoryMB(
  profile: CapabilityProfile,
  governor: GovernorReport
): number {
  if (governor.state === "paused" || governor.capacity <= 0) return 0;
  const vram = profile.gpu ? profile.gpu.vram * profile.gpu.available * governor.capacity : 0;
  // Always leave the owner headroom so their session never swaps.
  const ramFloor = 1024;
  const ram = Math.max(0, profile.memory.available * governor.capacity - ramFloor);
  const usable = vram > 0 ? vram + 0.12 * ram : 0.35 * ram;
  return Math.min(usable, governor.maxModelMemoryMB || usable);
}

/**
 * Capability-relative throughput prior, used before a node has produced any
 * measured tokens/second. Deliberately vendor-neutral: it reads the profile,
 * not a table of GPU names.
 */
export function throughputPrior(
  profile: CapabilityProfile,
  governor: GovernorReport
): number {
  if (governor.state === "paused" || governor.capacity <= 0) return 0;
  let score = 0;
  if (profile.gpu && profile.gpu.backend !== "cpu" && profile.gpu.backend !== "none") {
    // VRAM size correlates with class of accelerator far better than any name match.
    const memFactor = Math.sqrt(Math.max(profile.gpu.vram, 512) / 1024);
    const backendFactor =
      profile.gpu.backend === "cuda"
        ? 1.0
        : profile.gpu.backend === "metal"
          ? 0.85
          : profile.gpu.backend === "vulkan"
            ? 0.7
            : 0.4; // webgpu in a browser tab is the least predictable
    score += 12 * memFactor * backendFactor * profile.gpu.available;
  }
  score += 0.25 * profile.cpu.cores * profile.cpu.available;
  return score * governor.capacity;
}

File: crates/community-protocol/src/capability.rs
use crate::task::ModelAdvertisement;
use community_core::NodeId;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum NodeKind {
    DesktopWorker,
    BrowserContributor,
    Client,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum UserActivity {
    Idle,
    Light,
    Active,
    Busy,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ThermalState {
    Normal,
    Warm,
    Hot,
    Critical,
    Unknown,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum AcceleratorBackend {
    Cuda,
    Vulkan,
    Metal,
    Webgpu,
    Cpu,
    None,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CpuProfile {
    pub model: String,
    pub cores: usize,
    pub available_fraction: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GpuProfile {
    pub vendor: String,
    pub model: String,
    pub vram_mb: usize,
    pub available_fraction: f32,
    pub backend: AcceleratorBackend,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryProfile {
    pub total_mb: usize,
    pub available_mb: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetworkProfile {
    pub latency_ms: f32,
    pub bandwidth_mbps: f32,
    pub jitter_ms: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserState {
    pub activity: UserActivity,
    pub thermal_state: ThermalState,
    pub on_battery: bool,
    pub battery_pct: Option<f32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RpcProfile {
    pub endpoint: Option<String>,
    pub offered_memory_mb: usize,
    pub can_head: bool,
    pub build_tag: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CapabilityProfile {
    pub node_id: NodeId,
    pub label: String,
    pub kind: NodeKind,
    pub os: String,
    pub arch: String,
    pub cpu: CpuProfile,
    pub gpu: Option<GpuProfile>,
    pub memory: MemoryProfile,
    pub network: NetworkProfile,
    pub user_state: UserState,
    pub rpc: Option<RpcProfile>,
    pub cached_shards: Vec<String>,
    /// Models this peer can actually serve. Empty unless `state == ready`.
    #[serde(default)]
    pub models: Vec<ModelAdvertisement>,
    /// Whether this peer currently accepts *new* compute assignments.
    /// Missing on old peers → treated as true (they did not have a pause control).
    #[serde(default = "default_compute_sharing_enabled")]
    pub compute_sharing_enabled: bool,
    /// Extensibility for future shard/layer ranges. Empty means unspecified.
    #[serde(default)]
    pub supported_shard_ranges: Vec<String>,
}

fn default_compute_sharing_enabled() -> bool {
    true
}

/// Local opt-in compute contribution. Default is PAUSED (must be explicit).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct ResourceSharingConfig {
    pub enabled: bool,
    pub cpu_limit_percent: u32,
    pub memory_limit_mb: u64,
    pub gpu_enabled: bool,
    pub gpu_limit_percent: Option<u32>,
    pub idle_only: bool,
}

impl Default for ResourceSharingConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            cpu_limit_percent: 50,
            memory_limit_mb: 2048,
            gpu_enabled: false,
            gpu_limit_percent: None,
            idle_only: true,
        }
    }
}

impl ResourceSharingConfig {
    pub fn state_label(&self) -> &'static str {
        if self.enabled {
            "ACTIVE"
        } else {
            "PAUSED"
        }
    }

    pub fn clamp(mut self) -> Self {
        self.cpu_limit_percent = self.cpu_limit_percent.min(100);
        if let Some(g) = self.gpu_limit_percent {
            self.gpu_limit_percent = Some(g.min(100));
        }
        if !self.gpu_enabled {
            self.gpu_limit_percent = None;
        }
        self
    }
}
